File: php/inc/post.php
<?php
/**
 * Mailtexte und Versand.
 *
 * Zwei Faelle brauchen Post: die **Bestaetigung** einer neuen Registrierung
 * und das **Zuruecksetzen des Kennworts**. Beide laufen ueber einen
 * sechsstelligen Code, der eine halbe Stunde gilt.
 *
 * Versandwege (in der Konfiguration unter 'mail'):
 *   'mail'  - ueber die PHP-Funktion mail()
 *   'datei' - als Textdatei in daten/postfach (zum Ausprobieren und Testen)
 */

declare(strict_types=1);

const CODE_MINUTEN = 30;
const CODEDAUER    = CODE_MINUTEN * 60;
const GRUSS        = "Viele Gruesse\nKi Tu - Stundenplaner fuer das Kinderturnen";

function anrede(string $name): string
{
    $name = trim($name);
    return $name !== '' ? "Hallo $name," : 'Hallo,';
}

/** 123456 -> "123 456" - so liest es sich vom Bildschirm ab. */
function code_gruppiert(string $code): string
{
    return strlen($code) === 6 ? substr($code, 0, 3) . ' ' . substr($code, 3) : $code;
}

/** [Betreff, Text] fuer die Bestaetigung einer neuen Registrierung. */
function text_bestaetigung(string $name, string $code, string $adresse = ''): array
{
    $verweis = $adresse !== ''
        ? "\n" . rtrim($adresse, '/') . '/bestaetigen.php'
        : ' auf der Bestaetigungsseite.';
    $text = anrede($name) . "\n\n"
        . "schoen, dass du beim Ki-Tu-Stundenplaner dabei bist. Mit diesem Code\n"
        . "schaltest du dein Konto frei:\n\n"
        . '    ' . code_gruppiert($code) . "\n\n"
        . 'Der Code gilt ' . CODE_MINUTEN . " Minuten. Gib ihn auf der Seite ein, die nach der\n"
        . "Registrierung offen ist - oder hier:" . $verweis . "\n\n"
        . "Danach kann es losgehen: Ort und Gruppe waehlen, planen, Stundenbild als PDF\n"
        . "speichern.\n\n"
        . "Hast du dich nicht registriert? Dann ist diese Mail hinfaellig - ohne den\n"
        . "Code passiert nichts, und das Konto verfaellt von selbst.\n\n"
        . GRUSS . "\n";
    return ['Dein Bestaetigungscode fuer den Ki-Tu-Stundenplaner', $text];
}

/** [Betreff, Text] fuer ein vergessenes Kennwort. */
function text_kennwort(string $name, string $code, string $adresse = ''): array
{
    $verweis = $adresse !== ''
        ? "\n" . rtrim($adresse, '/') . '/kennwort-neu.php'
        : " auf der Seite 'Kennwort neu'.";
    $text = anrede($name) . "\n\n"
        . "du moechtest dein Kennwort fuer den Ki-Tu-Stundenplaner neu setzen. Dieser\n"
        . "Code macht den Weg frei:\n\n"
        . '    ' . code_gruppiert($code) . "\n\n"
        . 'Der Code gilt ' . CODE_MINUTEN . " Minuten und laesst sich nur einmal verwenden.\n"
        . "Gib ihn zusammen mit deinem neuen Kennwort hier ein:" . $verweis . "\n\n"
        . "Kam die Anfrage nicht von dir? Dann ignoriere diese Mail einfach - dein\n"
        . "bisheriges Kennwort bleibt unveraendert gueltig.\n\n"
        . GRUSS . "\n";
    return ['Neues Kennwort fuer den Ki-Tu-Stundenplaner', $text];
}

/** Verschickt die Mail - oder legt sie ins Postfach. */
function mail_senden(string $an, string $betreff, string $text): bool
{
    $konfig = konfig();
    $absender = $konfig['absender'] ?? 'Ki Tu <kitu@localhost>';

    if (($konfig['mail'] ?? 'mail') === 'datei') {
        $ordner = $konfig['daten'] . '/postfach';
        if (!is_dir($ordner)) {
            mkdir($ordner, 0770, true);
        }
        // Der Name muss streng aufsteigen, damit "die letzte Mail" eindeutig ist.
        $zeit = microtime(true);
        $name = gmdate('Ymd-His', (int) $zeit)
            . sprintf('-%06d', (int) round(($zeit - floor($zeit)) * 1000000))
            . '_' . preg_replace('/[^a-z0-9._-]+/', '_', mb_strtolower($an)) . '.txt';
        file_put_contents(
            $ordner . '/' . $name,
            "An: $an\nVon: $absender\nBetreff: $betreff\n\n$text"
        );
        return true;
    }

    $kopf = "From: $absender\r\n"
        . "Content-Type: text/plain; charset=utf-8\r\n"
        . "Content-Transfer-Encoding: 8bit\r\n";
    return mail($an, $betreff, $text, $kopf);
}

/** Der sechsstellige Code aus einem Mailtext - fuer Tests und Fehlersuche. */
function code_aus_text(string $text): string
{
    return preg_match('/^ {2,}(\d{3}) (\d{3})\s*$/m', $text, $treffer)
        ? $treffer[1] . $treffer[2]
        : '';
}

File: werkzeuge/post.py
#!/usr/bin/env python3
"""Mailversand und die Texte dazu.

Zwei Faelle brauchen Post: die **Bestaetigung** einer neuen Registrierung und
das **Zuruecksetzen des Kennworts**. Beide laufen ueber einen sechsstelligen
Code, der eine halbe Stunde gilt.

Versandwege:

* ``Dateipost`` legt jede Mail als Textdatei in ``server/postfach/`` ab. Das
  ist die Vorgabe - so laesst sich alles ohne Mailserver ausprobieren und
  pruefen.
* ``SMTPPost`` verschickt ueber einen richtigen Mailserver (STARTTLS, Login).

Der Server waehlt ueber ``--smtp`` zwischen beiden.
"""

from __future__ import annotations

import re
import smtplib
import ssl
from datetime import datetime, timezone
from email.message import EmailMessage
from pathlib import Path

ABSENDER = "Ki Tu - Stundenplaner <kitu@localhost>"
GRUSS = "Viele Gruesse\nKi Tu - Stundenplaner fuer das Kinderturnen"
CODE_MINUTEN = 30


def _anrede(name: str) -> str:
    return f"Hallo {name.strip()}," if name and name.strip() else "Hallo,"


def _gruppiert(code: str) -> str:
    """123456 -> '123 456' - so liest es sich vom Bildschirm ab."""
    return f"{code[:3]} {code[3:]}" if len(code) == 6 else code


def text_bestaetigung(name: str, code: str, adresse: str = "") -> tuple:
    """(Betreff, Text) fuer die Bestaetigung einer neuen Registrierung."""
    verweis = f"\n{adresse.rstrip('/')}/bestaetigen" if adresse else ""
    text = f"""{_anrede(name)}

schoen, dass du beim Ki-Tu-Stundenplaner dabei bist. Mit diesem Code
schaltest du dein Konto frei:

    {_gruppiert(code)}

Der Code gilt {CODE_MINUTEN} Minuten. Gib ihn auf der Seite ein, die nach der
Registrierung offen ist - oder hier:{verweis or " auf der Bestaetigungsseite."}

Danach kann es losgehen: Ort und Gruppe waehlen, planen, Stundenbild als PDF
speichern.

Hast du dich nicht registriert? Dann ist diese Mail hinfaellig - ohne den
Code passiert nichts, und das Konto verfaellt von selbst.

{GRUSS}
"""
    return "Dein Bestaetigungscode fuer den Ki-Tu-Stundenplaner", text


def text_kennwort(name: str, code: str, adresse: str = "") -> tuple:
    """(Betreff, Text) fuer ein vergessenes Kennwort."""
    verweis = f"\n{adresse.rstrip('/')}/kennwort-neu" if adresse else ""
    text = f"""{_anrede(name)}

du moechtest dein Kennwort fuer den Ki-Tu-Stundenplaner neu setzen. Dieser
Code macht den Weg frei:

    {_gruppiert(code)}

Der Code gilt {CODE_MINUTEN} Minuten und laesst sich nur einmal verwenden.
Gib ihn zusammen mit deinem neuen Kennwort hier ein:{verweis or " auf der Seite 'Kennwort neu'."}

Kam die Anfrage nicht von dir? Dann ignoriere diese Mail einfach - dein
bisheriges Kennwort bleibt unveraendert gueltig.

{GRUSS}
"""
    return "Neues Kennwort fuer den Ki-Tu-Stundenplaner", text


class Postausgang:
    """Gemeinsame Schnittstelle: eine Mail verschicken."""

    def sende(self, an: str, betreff: str, text: str) -> None:  # pragma: no cover
        raise NotImplementedError


class Dateipost(Postausgang):
    """Legt jede Mail als lesbare Textdatei ab - fuer Probelauf und Tests."""

    def __init__(self, ordner: Path, absender: str = ABSENDER) -> None:
        self.ordner = Path(ordner)
        self.ordner.mkdir(parents=True, exist_ok=True)
        self.absender = absender

    def sende(self, an: str, betreff: str, text: str) -> None:
        zeit = datetime.now(timezone.utc).strftime("%Y%m%d-%H%M%S-%f")
        sauber = re.sub(r"[^a-z0-9._-]+", "_", an.lower())
        datei = self.ordner / f"{zeit}_{sauber}.txt"
        datei.write_text(
            f"An: {an}\nVon: {self.absender}\nBetreff: {betreff}\n\n{text}",
            encoding="utf-8",
        )

    def letzte(self, an: str = None) -> str:
        """Zuletzt abgelegte Mail (an diese Adresse) - fuer die Tests."""
        dateien = sorted(self.ordner.glob("*.txt"))
        if an:
            sauber = re.sub(r"[^a-z0-9._-]+", "_", an.lower())
            dateien = [d for d in dateien if d.name.endswith(f"_{sauber}.txt")]
        return dateien[-1].read_text(encoding="utf-8") if dateien else ""


class SMTPPost(Postausgang):
    """Versand ueber einen richtigen Mailserver."""

    def __init__(self, wirt: str, port: int = 587, nutzer: str = "",
                 kennwort: str = "", absender: str = ABSENDER, tls: bool = True) -> None:
        self.wirt = wirt
        self.port = port
        self.nutzer = nutzer
        self.kennwort = kennwort
        self.absender = absender
        self.tls = tls

    def sende(self, an: str, betreff: str, text: str) -> None:  # pragma: no cover
        nachricht = EmailMessage()
        nachricht["From"] = self.absender
        nachricht["To"] = an
        nachricht["Subject"] = betreff
        nachricht.set_content(text)
        with smtplib.SMTP(self.wirt, self.port, timeout=20) as verbindung:
            if self.tls:
                verbindung.starttls(context=ssl.create_default_context())
            if self.nutzer:
                verbindung.login(self.nutzer, self.kennwort)
            verbindung.send_message(nachricht)


def code_aus_text(text: str) -> str:
    """Der sechsstellige Code aus einer Mail - fuer Tests und Fehlersuche."""
    treffer = re.search(r"^\s{2,}(\d{3}) (\d{3})\s*$", text, re.MULTILINE)
    return treffer.group(1) + treffer.group(2) if treffer else ""
